libraries/AP_InertialSensor/AP_InertialSensor_tempcal.cpp

```diff
@@ -0,0 +1,512 @@
+    // @Calibration: 1
+
+    // @Param: GYR1_Y
+    // @DisplayName: Gyroscope 1st order temperature coefficient Y axis
+    // @Description: This is the 1st order temperature coefficient from a temperature calibration
+    // @User: Advanced
+    // @Calibration: 1
+
+    // @Param: GYR1_Z
+    // @DisplayName: Gyroscope 1st order temperature coefficient Z axis
+    // @Description: This is the 1st order temperature coefficient from a temperature calibration
+    // @User: Advanced
+    // @Calibration: 1
+    
+    AP_GROUPINFO("GYR1", 7, AP_InertialSensor::TCal, gyro_coeff[0], 0),
+
+    // @Param: GYR2_X
+    // @DisplayName: Gyroscope 2nd order temperature coefficient X axis
+    // @Description: This is the 2nd order temperature coefficient from a temperature calibration
+    // @User: Advanced
+    // @Calibration: 1
+
+    // @Param: GYR2_Y
+    // @DisplayName: Gyroscope 2nd order temperature coefficient Y axis
+    // @Description: This is the 2nd order temperature coefficient from a temperature calibration
+    // @User: Advanced
+    // @Calibration: 1
+
+    // @Param: GYR2_Z
+    // @DisplayName: Gyroscope 2nd order temperature coefficient Z axis
+    // @Description: This is the 2nd order temperature coefficient from a temperature calibration
+    // @User: Advanced
+    // @Calibration: 1
+
+    AP_GROUPINFO("GYR2", 8, AP_InertialSensor::TCal, gyro_coeff[1], 0),
+
+    // @Param: GYR3_X
+    // @DisplayName: Gyroscope 3rd order temperature coefficient X axis
+    // @Description: This is the 3rd order temperature coefficient from a temperature calibration
+    // @User: Advanced
+    // @Calibration: 1
+
+    // @Param: GYR3_Y
+    // @DisplayName: Gyroscope 3rd order temperature coefficient Y axis
+    // @Description: This is the 3rd order temperature coefficient from a temperature calibration
+    // @User: Advanced
+    // @Calibration: 1
+
+    // @Param: GYR3_Z
+    // @DisplayName: Gyroscope 3rd order temperature coefficient Z axis
+    // @Description: This is the 3rd order temperature coefficient from a temperature calibration
+    // @User: Advanced
+    // @Calibration: 1
+
+    AP_GROUPINFO("GYR3", 9, AP_InertialSensor::TCal, gyro_coeff[2], 0),
+
+    AP_GROUPEND
+};
+
+
+/*
+  evaluate a 3rd order polynomial (without the constant term) given a set of coefficients
+ */
+Vector3f AP_InertialSensor::TCal::polynomial_eval(float tdiff, const AP_Vector3f coeff[3]) const
+{
+    // evaluate order 3 polynomial
+    const Vector3f *c = (Vector3f *)&coeff[0];
+    return (c[0] + (c[1] + c[2]*tdiff)*tdiff)*tdiff*INV_SCALE_FACTOR;
+}
+
+/*
+  correct a single sensor for the current temperature
+ */
+void AP_InertialSensor::TCal::correct_sensor(float temperature, float cal_temp, const AP_Vector3f coeff[3], Vector3f &v) const
+{
+    if (enable != Enable::Enabled) {
+        return;
+    }
+    temperature = constrain_float(temperature, temp_min, temp_max);
+    cal_temp = constrain_float(cal_temp, temp_min, temp_max);
```

Comment:
This calibration is a nice improvement Tridge !

For the extrapolation outside the calibration temperature range, another way instead of constrain is to compute the slope at extremities and do a linear extrapolation, like :
![thermal_bias_calib_extrapolation](https://user-images.githubusercontent.com/12662109/104843974-b5688d00-58cd-11eb-9a7f-56cfd4e1d69f.png)

this better matches the real offset behavior of accelero/gyros (physical thermal expansion does not stop at temp_min/max ;-) ) while avoiding the divergence of the 3rd order.
The slope at temp_min and temp_max can be obtained easily as the polynomial derivative is just dropping the c0 term :
poly_extrap_above(T°) = poly(temp_max - TEMP_REF) + (T° - temp_max) * slope(temp_max - TEMP_REF)
slope(tdiff) = (2*c[1] + 3*c[2]*tdiff)*tdiff*INV_SCALE_FACTOR;

what do you think?

Reply:
For the specific case that you mentioned here this procedure will hold. But at the same time, there can be many scenarios where instead of making it good this policy could make the situation worse. For instance in your plot if you are to choose the Temp min some measure after maxima and temp_max some measures before minima the 1st order extrapolation is actually going to worsen the situation as it will take the result in the opposite direction that the actual direction the curve is going to take.
constraining doesn't speculate which direction the curve is going to take and a very simple measure to take, and more importantly ensures that we have fairly repeatable behaviour outside the range.

Reply:
this may be worth considering in the future (maybe as an INS_TCAL_OPTIONS bit?). I want to avoid scope creep for the initial PR, so won't implement now